### PR TITLE
KAFKA-5311: Support ExtendedDeserializer in Kafka Streams.

### PR DESCRIPTION
**Pull request by subnova**
The contribution is my original work and I license the work to the project under the project's open source license.

This is a clone of the previous PR #3125 but now against trunk.

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        RecordHeaders headers = new RecordHeaders();
+        String deserializeKey = sourceNode.deserializeKey("topic", headers, "data".getBytes(StandardCharsets.UTF_8));
+        assertThat(deserializeKey, is("topic" + headers + "data"));
+    }
+
+    @Test
+    public void shouldProvideTopicHeadersAndDataToValueDeserializer() {
+        SourceNode<String, String> sourceNode = new MockSourceNode<>(new String[]{""}, new TheExtendedDeserializer(), new TheExtendedDeserializer());
+        RecordHeaders headers = new RecordHeaders();
+        String deserializeKey = sourceNode.deserializeValue("topic", headers, "data".getBytes(StandardCharsets.UTF_8));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java]: Nit: var should be named `deserializeValue`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.kafka.test.MockSourceNode;
+import org.junit.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class SourceNodeTest {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java:31]: Nit: please add `final` to all local vars and method parameters

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    public static class TheExtendedDeserializer implements ExtendedDeserializer<String> {
+        @Override
+        public String deserialize(String topic, Headers headers, byte[] data) {
+            return topic + headers + new String(data, StandardCharsets.UTF_8);
+        }
+
+        @Override
+        public void configure(Map<String, ?> configs, boolean isKey) {
+
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java]: Nit: get rid of empty line: `public void configure(Map<String, ?> configs, boolean isKey) { }`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        @Override
+        public String deserialize(String topic, byte[] data) {
+            return deserialize(topic, null, data);
+        }
+
+        @Override
+        public void close() {
+
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SourceNodeTest.java]: as above

**Review by mjsax — APPROVED** [review]
Thanks for adding a test. Just some nits.

**Comment by ijuma** [comment]
@dguy, is this OK?

**Review by dguy — APPROVED** [review]
Thanks @subnova LGTM

**Review by ijuma — APPROVED** [review]
Thanks for the PR, LGTM. Merging to trunk and 0.11.0. Since it's late in the cycle, please take a look @guozhangwang. We can do a follow-up if necessary.

**Comment by guozhangwang** [comment]
LGTM! Thanks @subnova

**Pull request closed without merging**